gateway/ln-gateway/src/lnd.rs

```diff
@@ -0,0 +1,302 @@
+use std::collections::HashMap;
+use std::fmt;
+use std::sync::Arc;
+
+use anyhow::anyhow;
+use async_trait::async_trait;
+use bitcoin_hashes::{sha256, Hash};
+use fedimint_core::task::TaskGroup;
+use secp256k1::PublicKey;
+use tokio::sync::{mpsc, Mutex};
+use tokio_stream::wrappers::ReceiverStream;
+use tonic_lnd::lnrpc::{GetInfoRequest, SendRequest};
+use tonic_lnd::routerrpc::{CircuitKey, ForwardHtlcInterceptResponse};
+use tonic_lnd::{connect, LndClient};
+use tracing::{error, info, trace};
+
+use crate::gatewaylnrpc::complete_htlcs_request::{Action, Cancel, Settle};
+use crate::gatewaylnrpc::get_route_hints_response::RouteHint;
+use crate::gatewaylnrpc::{
+    CompleteHtlcsRequest, CompleteHtlcsResponse, GetPubKeyResponse, GetRouteHintsResponse,
+    PayInvoiceRequest, PayInvoiceResponse, SubscribeInterceptHtlcsRequest,
+    SubscribeInterceptHtlcsResponse,
+};
+use crate::lnrpc_client::{HtlcStream, ILnRpcClient};
+use crate::GatewayError;
+
+type OutcomeMap = Arc<Mutex<HashMap<sha256::Hash, (LndSenderRef, Option<CircuitKey>)>>>;
+
+pub struct GatewayLndClient {
+    client: LndClient,
+    outcomes: OutcomeMap,
+    task_group: TaskGroup,
+}
+
+// Reference to a sender that forwards ForwardHtlcInterceptResponse messages to
+// LND
+type LndSenderRef = Arc<mpsc::Sender<ForwardHtlcInterceptResponse>>;
+
+impl GatewayLndClient {
+    pub async fn new(
+        address: String,
+        tls_cert: String,
+        macaroon: String,
+        task_group: TaskGroup,
+    ) -> crate::Result<Self> {
+        let client = connect(address, tls_cert, macaroon).await.map_err(|e| {
+            error!("Failed to connect to lnrpc server: {:?}", e);
+            GatewayError::Other(anyhow!("Failed to connect to lnrpc server"))
+        })?;
+
+        Ok(Self {
+            client,
+            outcomes: Arc::new(Mutex::new(HashMap::new())),
+            task_group,
+        })
+    }
+}
+
+impl fmt::Debug for GatewayLndClient {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "LndClient")
+    }
+}
+
+#[async_trait]
+impl ILnRpcClient for GatewayLndClient {
+    async fn pubkey(&self) -> crate::Result<GetPubKeyResponse> {
+        let mut client = self.client.clone();
+
+        let info = client
+            .lightning()
+            .get_info(GetInfoRequest {})
+            .await
+            .expect("failed to get info")
+            .into_inner();
+
+        let pub_key: PublicKey = info.identity_pubkey.parse().expect("invalid pubkey");
+        info!("fetched pubkey {:?}", pub_key);
+
+        Ok(GetPubKeyResponse {
+            pub_key: pub_key.serialize().to_vec(),
+        })
+    }
+
+    async fn routehints(&self) -> crate::Result<GetRouteHintsResponse> {
```

Comment:
Maybe add a comment that this isn't properly implemented yet

Reply:
tracked a TODO for issue #1953